### PR TITLE
WIP: Remove the use of NgModel (replace with Any where appropriate)

### PR DESCRIPTION
**Pull request by Bhashit**
As we discussed, I am opening this PR for further changes/discussion. I think there is still some work left here.

### CODE REVIEW

File: src/main/scala/net/liftmodules/ng/Angular.scala [hunk 1]
```diff
@@ -353,12 +353,12 @@ object Angular extends DispatchSnippet with AngularProperties with LiftNgJsHelpe
   /**
    * Transparently wraps `NgModel => Box[Any]` functions appropriately for `defModelToAny`/`jsonCall`
    */
-  implicit def wrapModelFnBox[M](f: M => Box[Any]) = ModelFnBox(f)
+  implicit def wrapModelFnBox[M](f: M => Box[Any]): ModelFnBox[M] = ModelFnBox(f)
```
Comment (Bhashit) [inline src/main/scala/net/liftmodules/ng/Angular.scala]: @joescii your suggestion about removing this wrappers. Should we keep them around as deprecated artifacts, and maybe remove them in a subsequent version?

**Comment by Bhashit** [comment]
@joescii I am not sure about a few things here. Now that the `defModelToFutureAny` can accept any type:

1. Do you think a better name might be needed? I mean, we can create a new method with that signature, while deprecating `defModelToFutureAny`. But I'm not sure what would be a better name. `defXToFutureAny`? What could X be?
2. Also, `defStringToFutureAny` can now be covered by `defModelToFutureAny`, so that should be deprecated as well?

File: src/main/scala/net/liftmodules/ng/Angular.scala (hunk 1 above)
Comment (joescii) [reply src/main/scala/net/liftmodules/ng/Angular.scala]: They can be removed entirely. They're an artifact of enforcing the `NgModel` constraint at compile time. Without the wrappers, the code would compile but fail at runtime. I testing removing these, and everything seems fine.

**Comment by joescii** [comment]
Yeah, this gives us an opportunity to think again about the function names. I've already confirmed that the functions accepting a `String` are now redundant and can be deprecated. I'll have to play with it to see if the specific function names are still needed. I created all of those to help with type inference, but now type inference is much less important. :) 

So in short, we just need to think it through. I'll play with the API some more to see what can be removed or merged together.

**Comment by joescii** [comment]
I'm finally getting back around to this PR. I feel this PR really comes down to getting names optimal...

With the removal of `NgModel`, `JsObjFactory` really only exposes 5 different type signatures for defining service functions:

```scala
=> Box[Any]
M => Box[Any]
=> Future[Box[Any]]
M => Future[Box[Any]]
Any
```

Note that in the last case, `Any` is a value calculated on page load (see `def anyVal`). The other 4 define functions which roundtrip from the client to the server at the time the function is invoked. This is why I opted for either `val` or `def` in method names to convey this difference using semantics familiar to a Scala developer. (I really have NO idea why I have `def defAny` etc, but `def anyVal` is reversed).

Regardless of past API name history, what are the five best names in your opinion? I feel they just need to describe the type. Perhaps something like...

```scala
def defAny(=> Box[Any]) // Same as today
def defArgToAny(M => Box[Any])
def defFutureAny(=> Future[Box[Any]]) // Same as today
def defArgToFutureAny(M => Future[Box[Any]])
def valAny(Any) // Reverse of today's name for consistency
```

All other existing methods would be deprecated and routed to these 5. What are your thoughts?

**Comment by Bhashit** [comment]
This seems pretty much alright to me. One thing that I got in my brain while looking at this was that maybe the term `arg` can be replaced with the term `param`. For ex. `defParamToFutureAny`. This is roughly based on [this idea](https://stackoverflow.com/a/156787/1258479). Also, it might make sense to say that it is a 'parameterized' method. Something like that. But I think this would definitely have to be your call.

**Comment by joescii** [comment]
I like `param` better too, especially in light of the SO post. That's a nice subtlety in the English language I hadn't noticed yet :)

**Comment by geoffreyyip** [comment]
Hey all, want to make sure of something. Would this change have any effects on Angular's directive `ng-model`?

**Comment by joescii** [comment]
Not at all! This is 100% on the server-side.

**Comment by joescii** [comment]
Take a look at the [remove-ng-model2](https://github.com/joescii/lift-ng/tree/remove-ng-model2) branch I created.

**Comment by Bhashit** [comment]
LGTM.

File: src/main/scala/net/liftmodules/ng/Angular.scala (hunk 1 above)
Comment (colinbes) [reply src/main/scala/net/liftmodules/ng/Angular.scala]: plus one for version numbering change, this sounds a significant and worthwhile change.

**Comment by colinbes** [comment]
Couple of comments for what it's worth:

I understand the allure of using **Any** in defStringToFutureAny etc, but are there hard reason to use **Any** as opposed to **AnyRef**? Wouldn't Long, Int etc all be passed as strings at the end of the day? Is 'object' model not safer? Guess all relative and pretty personal :-)

Likewise for comment using **Model** in functions like defModelToFutureAny, my personal preference is to use domain models where I can (Objects, think AnyRef) and not just pass types (besides strings). Just my preference. From client side angular calling service function a object (AnyRef, json object etc) just makes sense to me

On deprecated artifacts note, from someone who has recently had to go through a massive refactoring due to library changes I'd suggest rather dropping what's not needed and noting the breaking change. I find code just hanging around just helps create confusion.

**Comment by Bhashit** [comment]
FWIW, I'll try to reply to some of these comments.

> are there hard reason to use Any as opposed to AnyRef? Wouldn't Long, Int etc all be passed as strings at the end of the day? Is 'object' model not safer?

I think `Int`, `Long` etc would be passed on as JS numbers. I don't think they would be passed as Strings. Also, I don't know what 'safer' means in this context. But I don't think there is any safety issue involved there. As far as lift-json is concerned, it can deal with primitives just as well as the refs.

> I'd suggest rather dropping what's not needed and noting the breaking change

That might prevent people from gradually upgrading though. For ex. we have a whole bunch of code that relies on it. If we had to remove all those references, it would be a pretty huge task. Whereas deprecation would allow us to take care of it in parallel, gradually, while all the new code can start using the new APIs immediately.

**Comment by colinbes** [comment]
All good reasons!

**Comment by joescii** [comment]
So here's another thing we should consider at this point... We could use scala's `Future` rather than Lift's `LAFuture`. The reason I went with `LAFuture` in the first place was originally lift-ng supported versions of Scala before `Future` was introduced. As of this `0.11.0` release, I plan to only support Scala 2.11 and 2.12. 

Every lift-ng project I've seen has had to utilize the conversions from Scala's `Future` to `LAFuture`. So our two functions types which return a future could go with a Scala `Future` and I bet most everyone will be happy with that change. Thoughts?

**Comment by Bhashit** [comment]
Absolutely. That sounds perfect.

**Comment by joescii** [comment]
I got the API worked out where it now expects a `Future` rather than `LAFuture[Box]` (here)[https://github.com/joescii/lift-ng/blob/724cf744eba6956ce6bf4797124577022770c01f/src/main/scala/net/liftmodules/ng/Angular.scala#L363-L428]. Note that the code doesn't work. I just wanted to get the API defined and see what shakes out in projects I use it in. The changes seem manageable in large projects, including one which extends the `JsObjFactory` to override several methods.

I'm about to look into replacing lift-ng's internal use of `LAFuture` with `Future`. Again since I expect the overwhelming majority of lift-ng code to utilize scala's `Future`, I'd like to avoid the overhead of converting futures for the majority of use cases. Of course, the tradeoff here is that most users of lift-ng will be doing a double conversion (`Future` => `LAFuture` => `Future`) until they remove the deprecation warnings as plentiful as they may be.

**Comment by joescii** [comment]
I have everything worked out on [this branch](https://github.com/joescii/lift-ng/tree/v0.11.0). Any of you let me know if you want time to review before I publish in the next couple of days.

**Comment by joescii** [comment]
Closing the PR now that I've published the commits from here.

**Comment by joescii** [comment]
@colinbes Take a look at this new PR related to the work done here. #40

**Pull request closed without merging**